### DISCUSSION TITLE
Dynamic Dag Creation

### DISCUSSION THREAD

**Discussion by PankajKhanwani** [apache/airflow#33655] [Q&A]
I want to schedule something like a cron job according to user's payload. There can be any number of cron jobs and each of them can have different params. It is also not decided how many jobs will be there. Is there anyway I can achieve this with airflow??

**Comment by potiuk** [comment]
From that vague description - likely yes.

 Check out the documentation and examples and I recommend you to look through example dags  https://airflow.apache.org/docs/apache-airflow/stable/index.html 

You will find all the helpful documentation there when searching - with plenty of exampled.

Look for "Params", "Cron". 

There are also a lot of tutorials available and training. I heartily recommend you to pick somo of those mentioned in learning resources in https://airflow.apache.org/docs/apache-airflow/stable/index.html

You will find all the answers there.

**Comment by potiuk** [comment]
Specifically if you are looking for Dynamic DAGs and Dynamic Tasks you will find docs and useful examples here: https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#dynamic-dags
 
and here: https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html

**Comment by potiuk** [comment]
Good luck !

**Comment by PankajKhanwani** [comment]
So, I am writing an api which takes some arguments from the user about the cron expression. After user submit's request I am creating a new dag runtime which has its scheduled task according to the cron expression provided by the user.
I am facing 2 problems here:

1. In k8s, I am not able to do this as I was using gitsync. I changed to PVC but I need some pre-defined dags which will generate the user-required dags.

2. After sometime, there might me thousand of dags. Am I doing the right thing and also if Apache Airflow is the right way to go for this type of task. Or there is any other way around it

**Reply by potiuk** [reply]
Yes. It's OK to generate Python DAG files based on you some higher level abstraction. Many of our users do it.

1. You can still use GitSync, you just need to programmatically commit and push dags to your git (than airflow will sync) - it's even easier than to change them in the shared volume and you have all the history as well. I'd recommend you to use it.
3. Airflow can cope with thousands of Dags generally. It's up to you to manage them and do any kind of retention, cleanups etc. If you decide to create python DAG files programmatically, you need to take care about lifecycle of whatever your users will request. Nothing for Airflow - I assume you build your frontend application to manage it and it should take care about that.

**Comment by PankajKhanwani** [comment]
@potiuk I am using gitsync now and I am able to see dags in scheduler, triggerer and worker. But I am not able to view in Airflow UI.
